src/directory/managed_directory.rs

```diff
@@ -207,19 +210,198 @@ impl ManagedDirectory {
+        let tantivy_version = (footer[1], footer[2], footer[3]);
+        let crc = LittleEndian::read_u32(&footer[4..8]);
+
+        V0 {
+            tantivy_version,
+            crc,
+        }
+    }
+
+    pub fn from_crc(crc: u32) -> Self {
+        Self {
+            tantivy_version: (
+                env!("CARGO_PKG_VERSION_MAJOR").parse().unwrap(),
+                env!("CARGO_PKG_VERSION_MINOR").parse().unwrap(),
+                env!("CARGO_PKG_VERSION_PATCH").parse().unwrap(),
+            ),
+            crc,
+        }
+    }
+    pub fn size() -> u16 {
+        10
+    }
+}
+
+struct FooterProxy<W: TerminatingWrite> {
+    /// always Some except after terminate call
+    hasher: Option<Hasher>,
+    /// always Some except after terminate call
+    writer: Option<W>,
+}
+
+impl<W: TerminatingWrite> FooterProxy<W> {
+    fn new(writer: W) -> Self {
+        FooterProxy {
+            hasher: Some(Hasher::new()),
+            writer: Some(writer),
+        }
+    }
+}
+
+impl<W: TerminatingWrite> Write for FooterProxy<W> {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        let count = self.writer.as_mut().unwrap().write(buf)?;
+        self.hasher.as_mut().unwrap().update(&buf[..count]);
+        Ok(count)
+    }
+
+    fn flush(&mut self) -> io::Result<()> {
+        self.writer.as_mut().unwrap().flush()
+    }
+}
+
+impl<W: TerminatingWrite> TerminatingWrite for FooterProxy<W> {
+    fn terminate_ref(&mut self, _: AntiCallToken) -> io::Result<()> {
+        let crc = self.hasher.take().unwrap().finalize();
+        let footer = Footer::V0(V0::from_crc(crc)).to_bytes();
+        let mut writer = self.writer.take().unwrap();
+        writer.write_all(&footer)?;
+        writer.flush()?; //should we assume calling terminate on inner already flush?
+        writer.terminate()
+    }
 }
 
 impl Directory for ManagedDirectory {
     fn open_read(&self, path: &Path) -> result::Result<ReadOnlySource, OpenReadError> {
-        self.directory.open_read(path)
+        let read_only_source = self.directory.open_read(path)?;
+        let footer_size = Footer::from_bytes(read_only_source.as_slice()).size();
+        Ok(read_only_source.slice_to(read_only_source.as_slice().len() - footer_size))
     }
 
     fn open_write(&mut self, path: &Path) -> result::Result<WritePtr, OpenWriteError> {
         self.register_file_as_managed(path)
             .map_err(|e| IOError::with_path(path.to_owned(), e))?;
-        self.directory.open_write(path)
+        Ok(io::BufWriter::new(Box::new(FooterProxy::new(
+            self.directory
+                .open_write(path)?
+                .into_inner()
+                .map_err(|_| ())
+                .expect("buffer should be empty"),
+        ))))
     }
 
+    // TODO: to be correct, this require an non enforced contract, atomic writes must only be
+    // read back as atomics, and normal write must only be read by normal read
```

Comment:
I'm happy with this. Can you remove this comment and document this contract in `Directory` if it is not already there?